docs/specs/captcha.md

```diff
@@ -1,41 +1,452 @@
+
+### Risk assessment and Captcha in Authentication Flow configuration
+
+Risk assessment and Captcha are supported in the following flow types:
+
+- `signup`
+- `promote`
+- `login`
+- `signup_login`
+- `reauth`
+- `account_recovery`
+
+Risk assessment and Captcha are supported only in the following step types:
+
+- `identify` in `signup`, `promote`, `login`, `signup_login`, and `account_recovery`.
+- `create_authenticator` in `signup` and `promote`.
+- `authenticate` in `login` and `reauth`.
+
+We can see that all supported step types have branches.
+
+To enable risk assessment and Captcha in a branch, add `risk_assessment` and `captcha` to the branch.
+
+The configuration is as follows:
+
+```
+risk_assessment:
+  enabled: true
+  provider:
+    alias: recaptchav3
+
+captcha:
+  mode: "always" # "never" | "always" | "risk_level_low" | "risk_level_medium" | "risk_level_high"
+  fail_open: true
+  provider:
+    alias: cloudflare
+```
+
+- `captcha.mode`: When Captcha is required.
+  - `never`: Captcha is never required. It is the default.
+  - `always`: Captcha is always required. Risk level is ignored.
+  - `risk_level_low`: Captcha is required when the risk level obtained by risk assessment is low. If risk assessment is not enabled, then it means `always`. If risk assessment is service unavailable, then it means `always`.
+  - `risk_level_medium`: Captcha is required when the risk level obtained by risk assessment is medium. If risk assessment is not enabled, then it means `always`. If risk assessment is service unavailable, then it means `always`.
+  - `risk_level_high`: Captcha is required when the risk level obtained by risk assessment is high. If risk assessment is not enabled, then it means `always`. If risk assessment is service unavailable, then it means `always`.
+- `captcha.fail_open`: If it is true, then if the Captcha provider is service unavailable, access is granted. It is false by default.
+- `captcha.provider.alias`: If `mode` is not `never`, then it is required. Specify the Captcha provider to be used in this branch.
+
+For example,
+
+```
+authentication_flow:
+  login_flows:
+  - name: default
+    steps:
+    - type: identify
+      one_of:
+      - identification: email
+        # Identify with email requires captcha.
+        captcha:
+          mode: "always"
+          provider:
+            alias: cloudflare
+    - type: authenticate
+      one_of:
+      - authentication: primary_password
+      - authentication: primary_oob_otp_email
+```
+
+### Behavior of generated flows
+
+Given `captcha.enabled=true` and `captcha.providers` is non-empty,
+
+1. All the branches of the first step (that is, the `identify` step, or the `authenticate` step in reauth flow) has `captcha.mode=always`.
+2. The first provider in `captcha.providers` is used as `captcha.provider.alias`
+
+In terms of UX, when Captcha is enabled and configured, every generated flow requires captcha at the beginning of the flow.
```

Comment:
Missing behavior of generated flows for risk_assessment

```md
Given `risk_assessment.enabled=true` and `risk_assessment.providers` is non-empty,
```

Reply:
OK. I will add it back.